modules/renter/hostdb/negotiate.go

```diff
@@ -319,3 +311,70 @@ func newRevision(rev types.FileContractRevision, pieceLen uint64, merkleRoot cry
 		NewUnlockHash: rev.NewUnlockHash,
 	}
 }
+
+// Renew negotiates a new contract for data already stored with a host. It
+// returns the ID of the new contract. This is a blocking call that performs
+// network I/O.
+func (hdb *HostDB) Renew(fcid types.FileContractID, newEndHeight types.BlockHeight) (types.FileContractID, error) {
+	hdb.mu.RLock()
+	height := hdb.blockHeight
+	hc, ok := hdb.contracts[fcid]
+	host, eok := hdb.allHosts[hc.IP] // or activeHosts?
+	hdb.mu.RUnlock()
+	if !ok {
+		return types.FileContractID{}, errors.New("no record of that contract")
+	} else if !eok {
+		return types.FileContractID{}, errors.New("no record of that host")
+	} else if newEndHeight < height {
+		return types.FileContractID{}, errors.New("cannot renew below current height")
+	}
+
+	renterCost := host.Price.Mul(types.NewCurrency64(hc.LastRevision.NewFileSize)).Mul(types.NewCurrency64(uint64(newEndHeight - height)))
+	renterCost = renterCost.MulFloat(1.05) // extra buffer to guarantee we won't run out of money during revision
+	payout := renterCost                   // no collateral
+
+	// create file contract
+	fc := types.FileContract{
+		FileSize:       hc.LastRevision.NewFileSize,
+		FileMerkleRoot: hc.LastRevision.NewFileMerkleRoot,
+		WindowStart:    newEndHeight,
+		WindowEnd:      newEndHeight + host.WindowSize,
+		Payout:         payout,
+		UnlockHash:     types.UnlockHash{}, // to be filled in by negotiateContract
+		// TODO: are these correct?
+		RevisionNumber:     0,
+		ValidProofOutputs:  hc.LastRevision.NewValidProofOutputs,
+		MissedProofOutputs: hc.LastRevision.NewMissedProofOutputs,
```

Comment:
oh yeah, not sure what I was thinking.

You'd want the price to cover (filesize)*(duration)*(price), which should match the previous payouts assuming:
 
+ duration is the same
+ price is the same
+ the outputs weren't used as payment channels to pay for downloads

**So, I would definitely not use the outputs in the previous file contract, as it will probably cause future headaches. It should be calculated again using the size, duration, and price.**

I was also thinking that after the first renew, the second renew should be 6 months instead of just 30 days. Maybe 6 months is too long, but right now because we renew with only 2 weeks left, we have two big issues:

1. users who don't run Sia for 15 days straight have a non-zero probability of losing files.
2. users are paying 50% overhead for their renewing (14 days out of 30 are overlapping)

If we jump it to 6 months, or even just 3 months, we can extend the 2 weeks to 4 or 6 weeks and reduce the 50% overhead to something like 15%.

OTOH, users are not super price sensitive, and corporations can be given special setups that reduce the overheads (because a corporation can expect to run Sia 95% of the time, making the two weeks overkill).

Maybe we should leave it for now.